### PR TITLE
fix(automations): resolve bare OTEL attributes in metadata.value trigger filters

### PR DESCRIPTION
**Pull request by sergioestebance**
## What broke

Automation triggers filtering on `metadata.value` with a **bare OTEL resource attribute** key (e.g. `service.name`) silently never fire. The filter matches in the UI preview count, so it looks correct, and the trigger stays quiet forever.

Affects every tenant on the platform filtering this way. Found via a customer report; confirmed broken for an extended period.

## Cause

Two code paths resolve the same filter and they disagree.

**ClickHouse** (what the UI preview counts) ORs three candidates for a `metadata.value` key — `filters/clickhouse/filter-conditions.ts:97-113`:

```
metadata.{key}            canonical, from SDK canonicalization
langwatch.metadata.{key}  legacy REST collector
{key}                     legacy bare OTEL resource attribute
```

**In-memory matching** (what actually fires the trigger) only ever saw the first two. The trace-side builder feeds the matcher `customMetadata`, which is produced by `extractCustomMetadata` → `resolveCustomMetadataKey` (`filters/triggerFilter.matcher.ts:595-650`). That function drops any key starting with a standard OTEL resource prefix — `BARE_KEY_EXCLUDED_PREFIXES` at `:605-623` includes `"service."`, returning null at `:645`.

So the third candidate never reached the matcher. Traced end to end on a real production trace: of 19 attributes, `service.name` is the one dropped. The trace-side gate then closes at `triggerFilter.matcher.ts:301` and the evaluation half is never even consulted.

Meanwhile the filter dropdown offers these keys — `filters/clickhouse/filter-definitions.ts:200-224` lists `mapKeys(ts.Attributes)` excluding only `langwatch.` and `gen_ai.`, so `service.name` is presented as a valid choice. The UI was right; the matcher was the broken half.

## Fix

**1 — read all three candidates, off the raw attributes.** Give `PreconditionTraceData` an optional `attributes` field carrying the raw trace-summary attributes, and have the `metadata.value` matcher read the same three keys the ClickHouse builder reads, by the same names.

`customMetadata` cannot supply them. `extractCustomMetadata` drops standard resource prefixes outright, and where several forms of one key are present it keeps only the highest-priority one (`triggerFilter.matcher.ts:660-677`, priority 3/2/1) — so a trace carrying both `metadata.env` and `langwatch.metadata.env` exposes one value there while ClickHouse matches either. It stays in the candidate list as the only source for callers that carry parsed metadata and no raw attributes.

**2 — OR the candidates instead of taking the first** (raised in review). ClickHouse ORs its three; the matcher returned the first by precedence. With `metadata.env = "a"` and a bare `env = "b"`, a filter on `["b"]` counted in the preview and never fired, and a present-but-blank canonical key blocked the bare candidate outright. Candidates are now deduped and every distinct one returned, so the engine can match any — it already treats a list as any-element-matches (`triggerFilter.matcher.ts:332`).

A single distinct candidate stays a plain string rather than a one-element list, so every existing consumer keeps reading the value and not its encoding. That shape is asserted directly, because a test that only checks whether a filter matched passes either way and pins nothing.

**3 — `matches_regex` tested the encoding, not the values** (raised in review, follow-on from 2). Handing that rule an array exposed it as the only rule of the four that ran `JSON.stringify` over a list instead of iterating it — `is`, `contains` and `not_contains` already go element-wise (`evaluations/preconditions.ts:103,125,147`). So an anchored pattern could never match **any** array-valued field: `metadata.labels`, `metadata.prompt_ids`, `spans.type`, `spans.model`, `topics.topics`, `topics.subtopics`. `^gpt-4` was tested against `["gpt-4o"]`, a string starting with `[`.

Now each element is tested and any match passes. The serialized array stays in the candidate list, so a pattern written against the old JSON-encoded form keeps working — the rule can only gain matches, never lose them. AND semantics between separate preconditions are unchanged.

For `metadata.value` this was latent, not live: the precondition builder carries `customMetadata` and no raw attributes, so the two-candidate array never formed there (#8172). The array-regex bug itself was live for the other six fields.

`BARE_KEY_EXCLUDED_PREFIXES` is untouched, so `customMetadata` keeps its current contents.

## Failing output (before)

```
FAIL  > matchesTriggerFilters > when filtering by metadata.value on a bare OTEL
       resource attribute > matches a service.name attribute hoisted onto the
       trace summary

AssertionError: expected false to be true // Object.is equality

Tests  1 failed | 2 passed | 95 skipped (98)
```

Verified against base by restoring both source files from `main` and re-running a decisive test built from a **real production filter blob and the verbatim 19-attribute map of a real trace**:

```
traceFilters: {"metadata.value":{"service·name":[...],"metadata·trace_type":[]}}
trace half matches: false
AssertionError: expected false to be true
Tests  1 failed (1)
```

## Passing output (after)

`src/server/filters`, `src/server/evaluations` and `src/server/app-layer/automations`, run together:

```
Test Files  55 passed (55)
     Tests  868 passed (868)
```

`pnpm typecheck:all` is clean, and the CI lint gate run the way CI runs it reports no new Biome violations (base 3333, head 3333).

Generated files must exist first — `pnpm start:prepare:files`. Without them these suites fail to import `~/generated/prisma/client` and the SDK entry point, which looks like a regression and is not one.

## Sweep

Three sites construct `PreconditionTraceData`:

| Site | Status |
|---|---|
| `filters/triggerFilter.matcher.ts:147` | **Fixed here** |
| `evaluations/preconditions.ts:302` (legacy collector) | Safe by absence — that shape has no raw OTEL attribute map |
| `evaluations/preconditions.ts:371` (event-sourcing command) | **Same bug, not fixed here** — #8172 |

Also checked: no logging or serialization of this data anywhere, so the new field widens no PII surface. It stays in memory for matching only.

## Deliberately NOT fixed

1. **Monitor/evaluation preconditions have the same bug.** A `metadata.value` precondition on a bare OTEL key still resolves to null there. Fixing it crosses a persisted event-sourcing command schema (`ExecuteEvaluationCommandData` carries `customMetadata` and no raw attributes) and needs in-flight-command handling. Different feature, separate change. Tracked in #8172.

2. **`service.*` stays in the filter dropdown.** Removing it would take a working feature away instead of repairing it, and ClickHouse resolves it correctly today.

3. **`BARE_KEY_EXCLUDED_PREFIXES` unchanged.** Deleting `"service."` was the other candidate fix. Rejected: the exclusion is deliberate and pinned by an existing test (`triggerFilter.matcher.unit.test.ts:1090`), and changing it would alter `customMetadata` for every tenant.

4. **No parity test binds the two resolvers.** The other in-memory/ClickHouse pair in the codebase is pinned field-by-field in `filter-to-clickhouse/__tests__/evaluate.parity.unit.test.ts`. This pair is not, so the next edit to `filter-conditions.ts` can drift silently again — which is exactly how the missing key form above survived. Building that harness is a test-infrastructure change wider than this fix.

5. **The two key lists still diverge.** The filter dropdown (`clickhouse/filter-definitions.ts:200-224`) and the matcher's exclusion list (`BARE_KEY_EXCLUDED_PREFIXES`) are separately hand-maintained. This fix bypasses both by falling back to raw attributes, so the whole class is closed behaviourally, but the lists are still two sources of truth. Anyone adding an exclusion can recreate the shape. Merging them is a design change, not a hotfix.

6. **The raw attribute lookups keep using the un-stripped key.** A review pass suggested switching them to the prefix-stripped key. Rejected: the ClickHouse builder (`filter-conditions.ts:97-113`) also builds its three attribute names from the un-stripped key, so for a saved key like `metadata·env` both sides resolve through their third candidate and stay in step. Stripping only on the in-memory side would make it read attributes ClickHouse does not read for that key — a trace with `metadata.env = "prod"` and a bare `env = "staging"` would fire a `staging` trigger the preview counted as zero. That is this bug pointed the other way. The two dead candidates each side carries for prefixed keys are wasted work, not a defect, and removing them has to happen on both sides in one commit.

7. **The key dropdown still hides `langwatch.`-prefixed attributes.** `filter-definitions.ts:216` excludes them, so a project whose metadata only ever landed in that form gets no option to select even though both the filter condition and the matcher would resolve it if the key were typed by hand. Separate gap, upstream of this one, and it changes what every user sees in the dropdown.

## Deploy note

Affected triggers start firing again on deploy, roughly 6/hour for the largest affected tenant. The missed backlog **cannot** replay — `evaluationAlertTriggerMatch.subscriber.ts:43` drops events older than one hour, and nothing was ever enqueued because the match never happened.

Affected customers should be told before this lands, not after.

## Summary by CodeRabbit

* **Bug Fixes**
  * Trigger filters now match values from OpenTelemetry resource attributes.
  * Metadata matching handles values consistently across supported metadata key formats and conflicting sources.
  * Empty or unresolved metadata keys no longer produce unintended matches.
  * Regex preconditions now evaluate individual values in array fields, including span models.
  * Legacy trace data continues to use its available metadata sources without resolving raw resource attributes.

* **Tests**
  * Added coverage for resource attributes, conflicting metadata values, array-based regex matching, and non-matching values.

### CODE REVIEW

File: platform/app/src/server/filters/precondition-matchers.ts [hunk 1]
```diff
@@ -99,7 +105,16 @@ export const PRECONDITION_FIELD_MATCHERS: Record<
...
+    if (!resolved) return null;
+    // Canonical `metadata.{key}` and legacy `langwatch.metadata.{key}` are
+    // already hoisted into customMetadata with their prefix stripped.
+    const hoisted = data.customMetadata?.[resolved];
+    if (hoisted != null) return hoisted;
+    // Bare OTEL attribute. extractCustomMetadata deliberately keeps standard
+    // resource prefixes (service., http., telemetry., …) out of
+    // customMetadata, so this candidate has to come from the raw attributes
+    // or the filter can never match what the ClickHouse preview counted.
+    return data.attributes?.[decoded] ?? null;
```
[the remark this answers is not in the document]
Comment (sergioestebance) [reply platform/app/src/server/filters/precondition-matchers.ts]: Valid. Fixed in 3c61118c16.

The matcher returned the first candidate by precedence while ClickHouse ORs all three, so `metadata.env = "a"` plus a bare `env = "b"` matched the preview on `["b"]` and never fired the trigger. The same precedence also let a present-but-blank canonical key block the bare fallback entirely.

Now both candidates come back when they are both present and disagree, and the engine matches either — it already treats an array as any-element-matches (`triggerFilter.matcher.ts:332`).

One correction to the suggested shape: not a plain array in every case. `matches_regex` runs `JSON.stringify` over an array (`evaluations/preconditions.ts`), so always returning a list makes `^prod` stop matching `production` — it tests against `["production"]`. That exact version was written first and the new regex test failed on it. A single candidate stays a plain string, and a guard test now pins that so the trade cannot be made silently later.

Raw `metadata.{key}` and `langwatch.metadata.{key}` are not checked separately against `attributes` because `extractCustomMetadata` already hoists both into `customMetadata` with the prefix stripped, so candidates 1 and 2 are covered there.

File: platform/app/src/server/filters/precondition-matchers.ts [hunk 2]
```diff
@@ -99,7 +105,24 @@ export const PRECONDITION_FIELD_MATCHERS: Record<
...
+    // customMetadata, so this candidate has to come from the raw attributes
+    // or the filter can never match what the ClickHouse preview counted.
+    const bare = data.attributes?.[decoded] ?? null;
+    // ClickHouse ORs its three candidates rather than taking the first
+    // (clickhouse/filter-conditions.ts), so when both are present and
+    // disagree, hand back both and let the caller match either. Staying a
+    // plain string in every other case keeps `matches_regex`, which
+    // JSON.stringifies an array, reading the value and not its encoding.
+    if (hoisted != null && bare != null && hoisted !== bare) {
+      return [hoisted, bare];
```
[the remark this answers is not in the document]
Comment (sergioestebance) [reply platform/app/src/server/filters/precondition-matchers.ts]: Correct, and wider than the case you found. Fixed in 4e44d58a32.

`matches_regex` was the only rule testing `JSON.stringify(fieldValue)` instead of the values — `is`, `contains` and `not_contains` already iterate array elements (`evaluations/preconditions.ts:103,125,147`). So every array-valued field reaching this rule was affected, not just a conflicting `metadata.value` pair: `metadata.labels`, `metadata.prompt_ids`, `spans.type`, `spans.model`, `topics.topics`, `topics.subtopics`. An anchored pattern could never match any of them, because the string under test started with `[`.

Now each element is tested and any match passes, the same any-element semantics the other three rules use. AND semantics between separate preconditions are untouched — this only changes how one rule resolves one field value.

The serialized array stays in the candidate list. A pattern someone wrote against the old JSON-encoded form keeps matching, so the rule can only gain matches, never lose them.

One thing worth noting on the original report: for `metadata.value` specifically this was latent rather than live. The precondition path builds its trace data from `buildPreconditionTraceDataFromCommand`, which carries `customMetadata` and no raw attributes, so `bare` is always null there and the array never formed (that gap is #8172). The array-regex bug itself was live for the other six fields.

Two tests added, both verified failing before the change: a conflicting canonical/bare pair where only the bare value satisfies the pattern, and an anchored pattern over a span-model list.

**Comment by sergioestebance** [comment]
## Ruthless review of this branch — two real findings, both fixed

Reviewed `main...HEAD`. Two findings were real and blocking, and both were mine. Fixed in `5045707c29` and `996fc168c8`. Everything below is recorded so a reviewer can judge the fixes rather than take them on trust.

---

### 1. The matcher read two of ClickHouse's three key forms — fixed in `5045707c29`

ClickHouse resolves one `metadata.value` key against three independent attribute keys and ORs them (`filters/clickhouse/filter-conditions.ts:104-112`):

- `metadata.{key}`
- `langwatch.metadata.{key}`
- `{key}`

The matcher was reading two: `customMetadata`, plus the raw bare attribute added earlier on this branch. `customMetadata` cannot stand in for the other two, because `extractCustomMetadata` (`filters/triggerFilter.matcher.ts:660-677`) **collapses every form of a key down to the highest-priority one**:

```ts
const currentPriority = priorities[resolved.customKey] ?? 0;
if (resolved.priority <= currentPriority) continue;
```

So a trace carrying both `metadata.env = "prod"` and `langwatch.metadata.env = "staging"` exposes only `"prod"` there. A filter on `staging` matches in ClickHouse, counts in the preview, and never fires the trigger — the same defect this branch exists to fix, one key form over.

The fix reads the three attribute keys by the names the ClickHouse builder uses, keeps `customMetadata` in the list as the only source for callers that carry parsed metadata and no raw attributes, dedupes, and returns a plain string for one distinct candidate or a list for several.

Found by a convention-conformance scan against the ClickHouse/in-memory parity the file declares for itself. Code-traced and then pinned by a test that fails without the change.

### 2. A guard test I said pinned the resolver shape pinned nothing — fixed in `996fc168c8`

Earlier in this thread I wrote that a guard test pins the resolver staying a plain string "so the trade cannot be made silently later". That was true when written and stopped being true in `4e44d58a32`, which made `matches_regex` test array elements as well as the encoding. After that commit the assertion passes whether the resolver returns `"production"` or `["production"]`.

Verified rather than argued — forcing the resolver to always return a list:

```
Test Files  2 passed (2)
     Tests  197 passed (197)
```

A test that cannot fail is worse than no test, because it reads as coverage. The real guard now asserts the return value directly, in the matcher's own colocated test file. Same probe against the new guard:

```
× returns the value as a plain string
× returns a plain string when only the bare attribute is present
× returns a plain string when both candidates agree
     Tests  3 failed | 102 passed (105)
```

### 3. A test title named a code path that does not run — fixed in `996fc168c8`

`falls back to the bare attribute when the canonical key is blank` — a blank canonical value is a present candidate, not a missing one, so it goes through the ordinary any-candidate match, not a fallback. Renamed.

---

### Checked and deliberately not changed

**`not_contains` over a two-candidate list inverts.** Executed probe on the real `evaluatePreconditions` path, same trace data, only `attributes` differing:

```
PROBE withAttrs=false withoutAttrs=true
```

This is unreachable today: both precondition builders set `attributes: null`, and fold-state data reaches only `matchesTriggerFilters`, which has no `not_contains`. It is also the correct reading once attributes do arrive — negating an OR of candidates is an AND of nots, which is what the array branch produces. Recording it because #8172 is the change that will wire attributes into that path, and it should land this deliberately rather than discover it.

**No parity test binds the in-memory matcher to the ClickHouse builder.** The other in-memory/ClickHouse pair in the codebase is pinned field-by-field in `filter-to-clickhouse/__tests__/evaluate.parity.unit.test.ts`. This pair is not, so the next edit to `filter-conditions.ts` can drift silently. Not fixed here — it is a test-infrastructure change wider than this PR, and it would be the third scope expansion on a branch that started as a one-line resolver fix.

### Verification

- `src/server/filters`, `src/server/evaluations`, `src/server/app-layer/automations` — **868 passed (868)**
- `pnpm typecheck:all` — clean
- CI lint gate run as CI runs it, against the merge base — **no new Biome violations** (base 3333, head 3333)

### Honesty note on method

The convention scan that found finding 1 ran as an independent agent. The two refutation probes above I ran myself, in this same context. That is weaker than an independent check: I formed the claims and then judged them. The executed output is real, but it is not independent, and the difference is worth knowing.

---

Blessed. 🙏 The in-memory matcher and the ClickHouse builder now count the same trace, so a filter the preview offers is a filter the trigger can fire on.

> "A false balance is an abomination to the Lord, but a just weight is his delight." — Proverbs 11:1

Traced both resolver paths, the collapse in `extractCustomMetadata`, and every rule that can receive a list. No `P0`/`P1` left open. Residual risk is the missing parity test, waived through and named above. This is my own PR, so the approval is a human's to give — I cannot and should not bless my own work past this point.

**Comment by sergioestebance** [comment]
### Review of the fix commits themselves

`5045707c29` rewrote the resolver, so it needed reading by something other than the commit that wrote it. No new findings. The one property worth proving rather than asserting: callers that carry parsed metadata and no raw attributes — monitor preconditions, the legacy collector path — must behave exactly as they did before.

Compared the new resolver against the previous commit's logic across every attributes-less shape (customMetadata `null` / `{}` / value / empty string / explicit null / unrelated key, crossed with the bare, `metadata·`-prefixed, `langwatch·metadata·`-prefixed, empty-after-strip and empty keys):

```
PROBE diffs=0 (none)
```

Bit-identical on all 30. The change is inert for those callers and only adds candidates where raw attributes exist, which is the trigger-dispatch path this PR is about.

Also checked on the new code and found clean: the candidate list can only produce distinct values already present on the trace, so it never exceeds ClickHouse's three; the `metadata.${decoded}` lookups mirror the ClickHouse builder's un-stripped `rawKey` exactly; and a key that is empty after prefix stripping still returns null before any lookup runs.

File: platform/app/src/server/filters/precondition-matchers.ts [hunk 3]
```diff
@@ -99,7 +105,33 @@ export const PRECONDITION_FIELD_MATCHERS: Record<
...
+    // extractCustomMetadata drops standard resource prefixes (service., http.,
+    // telemetry., …) entirely, and where several forms of one key are present
+    // it keeps only the highest-priority one — so a trace carrying both
+    // `metadata.env` and `langwatch.metadata.env` exposes one value there
+    // while ClickHouse matches either. It stays in the list as the only source
+    // for callers that carry parsed metadata and no raw attributes.
+    const candidates = [
+      data.attributes?.[`metadata.${decoded}`],
+      data.attributes?.[`langwatch.metadata.${decoded}`],
+      data.attributes?.[decoded],
```
[the remark this answers is not in the document]
Comment (sergioestebance) [reply platform/app/src/server/filters/precondition-matchers.ts:125]: Checked this against the ClickHouse builder it has to agree with, and I'm leaving the lookups on `decoded`.

`decoded` is deliberate: it mirrors `filter-conditions.ts:97-113`, which also builds its three attribute names from the un-stripped key (`rawKey = key.replaceAll("·", ".")`). For `key: "metadata·env"` ClickHouse queries `metadata.metadata.env`, `langwatch.metadata.metadata.env` and `metadata.env` — and its third slot matches the canonical attribute. The matcher's `data.attributes?.[decoded]` is that same third slot. So prefixed keys do resolve; two of the three candidates are simply dead for them, on both sides equally.

Switching to `resolved` would make the matcher read attributes ClickHouse does not read for that key, and that is a false fire rather than a fix:

- trace has `metadata.env = "prod"` and a bare `env = "staging"`
- saved filter: key `metadata·env`, value `staging`
- ClickHouse: no match — the preview counts zero
- today: no match, same as the preview
- with `resolved`: bare `env` becomes a candidate, the trigger fires on a trace the preview never counted

That is the same preview/trigger divergence this PR exists to close, in the opposite direction.

Two supporting facts on the suggested test cases:

- `metadata·env` is the normal spelling, not an edge case — the key dropdown selects raw attribute names with dots encoded (`filter-definitions.ts:198-222`), so Python-SDK data canonicalized to `metadata.env` arrives as `metadata·env`. It is already covered by the existing cases at `triggerFilter.matcher.unit.test.ts:163-177`.
- `langwatch·metadata·env` cannot be selected at all — that same query excludes keys starting with `langwatch.` (`filter-definitions.ts:216`).

The dead candidates are wasted work on both sides, and the real gap is upstream: a project whose metadata only ever landed as `langwatch.metadata.*` gets no dropdown option even though the filter and the matcher would both resolve it. Both are worth their own change, and neither should ride along here, because moving the matcher without moving the ClickHouse builder in the same commit is exactly what reintroduces drift.

**Review by langwatch-agent — COMMENTED** [review]
No actionable findings. Reviewed the three raw metadata candidates against the ClickHouse key construction, legacy customMetadata fallback, array matching, and added regression tests. Validation was static; I did not rerun the unit suite or execute a live ClickHouse parity test. The documented command-time raw-attribute gap remains outside this fix.

LangWatch-Review: verdict=clean sha=b7031cda0da3b17c5af783c74ee74b2ac20b5653 p0=0 p1=0 p2=0 p3=0
